### PR TITLE
Submitting and Verifying your Smart Contract to Etherscan (Part 3/4 of Hello World dApp Series)

### PR DESCRIPTION
**Pull request by nstrike2**
## Description

Added header tags for each of the titles as per request from @minimalsm. Please make sure to approve/merge [changes to Part 1/4](https://ethereum.org/en/developers/tutorials/hello-world-smart-contract/) and [Part 2/4](https://github.com/ethereum/ethereum-org-website/pull/4280) before approving/merging this one.

## Related Issue

Continuation of https://github.com/ethereum/ethereum-org-website/pull/4159, which has now been closed due to inaccessibility.

Related Issue: https://github.com/ethereum/ethereum-org-website/issues/4092

### CODE REVIEW

**Comment by nstrike2** [comment]
@minimalsm @corwintines @pettinarip @samajammin @wackerow Can we get this merged? If not, happy to change what needs to be changed to get this merged.

**Review by minimalsm — CHANGES_REQUESTED** [review]

**Comment by nstrike2** [comment]
Hey @minimalsm , finished the changes for Part 3/4.

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+And if you haven't already, you'll definitely need an Alchemy account to complete any of these tutorials. Sign up for a free account [here](https://alchemy.com/)!
+
+## Part 3: Publish your Smart Contract to Etherscan
+
+You did all the hard work of bringing your smart contract to life - now it's time to share it with the world!
+
+By verifying your smart contract on Etherscan, anyone can view your source code and interact with your smart contract. Let's get started!
+
+This guide assumes you already have an [Alchemy account](https://alchemy.com/?a=eth-org-etherscan), access to our [Dashboard](https://dashboard.alchemyapi.io/), have gone through our [Hello World Smart Contract](./hello-world-smart-contract-for-beginners), and also have completed our [Interacting with a Smart Contract](./executing-and-interacting-with-a-smart-contract) tutorial.
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:38]: This is a duplicate of the above

```suggestion
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+### Step 1: Generate an API Key on your Etherscan account {#step-1-generate-an-api-key-on-your-etherscan-account}
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:40]:
```suggestion
### Step 1: Generate an API Key on your Etherscan account {#generate-an-api-key-on-your-etherscan-account}
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+```javascript
+API_URL = "https://eth-ropsten.alchemyapi.io/v2/your-api-key"
+MNEMONIC = "your-metamask-seed-reference"
+PUBLIC_KEY = "your-public-account-address"
+PRIVATE_KEY = "your-private-account-address"
+ETHERSCAN_API_KEY = "your-etherscan-key" 
+```
+
+### Step 2: HardHat-deployed smart contracts {#step-2-hardhat-deployed-smart-contracts}
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:79]:
```suggestion
### Step 2: HardHat-deployed smart contracts {#hardhat-deployed-smart-contracts}
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+This is where our steps diverge for HardHat and Truffle deployed smart contracts, as they require different plugins. Skip to the _Step 3: Truffle-deployed smart contracts_ section if you deployed your contract using Truffle.
+
+#### Step 2.1 Install the [`hardhat-etherscan`](https://hardhat.org/plugins/nomiclabs-hardhat-etherscan.html) plugin {#step-2.1-install-the-hardhat-etherscan-plugin}
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:83]: We can't link in headers

```suggestion
#### Step 2.1 Install hardhat-etherscan {#install-hardhat-etherscan}
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+  },
+  etherscan: {
+    // Your API key for Etherscan
+    // Obtain one at https://etherscan.io/
+    apiKey: ETHERSCAN_API_KEY
+  }
+};
+```
+
+#### Step 2.2 Verify your smart contract on Etherscan! {#step-2.2-verify-your-smart-contract-on-etherscan}
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:121]:
```suggestion
#### Step 2.2 Verify your smart contract on Etherscan {#verify-your-smart-contract-on-etherscan}
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+  },
+  plugins: ['truffle-plugin-verify'],
+  api_keys: {
+    etherscan: process.env.ETHERSCAN_API_KEY
+  }
+
+};
+```
+
+#### Step 3.2 Verify your smart contract on Etherscan! {#step-3.2-verify-your-smart-contract-on-etherscan}
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:239]:
```suggestion
#### Step 3.2 Verify your smart contract on Etherscan {#truffle-verify-your-smart-contract-on-etherscan}
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+If all goes well, you should see the following message in your terminal.
+
+```text
+Verifying HelloWorld 
+Pass - Verified: https://ropsten.etherscan.io/address/<contract-address>#contracts
+Successfully verified 1 contract(s).
+```
+
+### Step 4: Check out your smart contract on Etherscan! {#step-4-check-out-your-smart-contract-on-etherscan!}
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:255]:
```suggestion
### Step 4: Check out your smart contract on Etherscan! {#check-out-your-smart-contract-on-etherscan!}
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+When you navigate to the link provided in your terminal, you should be able to see your smart contract code and ABI published on Etherscan!
+
+![](https://gblobscdn.gitbook.com/assets%2F-MB17w56kk7ZnRMWdqOL%2F-MReibSVpLlOCa_Si494%2F-MRewuf9NzToMPg5WIln%2Fimage.png?alt=media&token=ca99b37e-6847-4b7a-8fef-371783f2e536)
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:259]: We should host any images we include on ethereum.org so we're not relying on third parties for their availability ([Check out the style guide for an example](https://ethereum.org/en/contributing/style-guide/#images)). Could you upload this image in the local directory and reference it that way instead?

```suggestion
![](https://gblobscdn.gitbook.com/assets%2F-MB17w56kk7ZnRMWdqOL%2F-MReibSVpLlOCa_Si494%2F-MRewuf9NzToMPg5WIln%2Fimage.png?alt=media&token=ca99b37e-6847-4b7a-8fef-371783f2e536)
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+https://ropsten.etherscan.io/address/<contract-address>#contracts
+```
+
+Congrats! Your smart contract code should be on Etherescan! Check out [Step 4](#step-4-check-out-your-smart-contract-on-etherscan!) to see how to view your smart contract code!
+
+### Step 3: Truffle-deployed smart contracts {#step-3-truffle-deployed-smart-contracts}
+
+Skip this section if you deployed your smart contract with HardHat.
+
+#### Step 3.1: Install the `truffle-plugin-verify` plugin {#step-3.1-install-the-truffle-plugin-verify-plugin}
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:151]:
```suggestion
#### Step 3.1: Install the `truffle-plugin-verify` plugin {#install-the-truffle-plugin-verify-plugin}
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+for verification on Etherscan. Waiting for verification result...
+
+
+Successfully verified contract HelloWorld on Etherscan.
+https://ropsten.etherscan.io/address/<contract-address>#contracts
+```
+
+Congrats! Your smart contract code should be on Etherescan! Check out [Step 4](#step-4-check-out-your-smart-contract-on-etherscan!) to see how to view your smart contract code!
+
+### Step 3: Truffle-deployed smart contracts {#step-3-truffle-deployed-smart-contracts}
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:147]:
```suggestion
### Step 3: Truffle-deployed smart contracts {#truffle-deployed-smart-contracts}
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+    // Your API key for Etherscan
+    // Obtain one at https://etherscan.io/
+    apiKey: ETHERSCAN_API_KEY
+  }
+};
+```
+
+#### Step 2.2 Verify your smart contract on Etherscan! {#step-2.2-verify-your-smart-contract-on-etherscan}
+
+Make sure your files are properly saved \(especially if you're in VSCode\) and your `.env` variables are properly configured.
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:123]:
```suggestion
Make sure your files are properly saved (especially if you're in VSCode) and your `.env` variables are properly configured.
```

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+  ]
+skill: beginner
+lang: en
+sidebar: true
+published: 2021-10-06
+---
+
+# Submitting your Smart Contract to Etherscan
+
+You should have completed part 1 [creating and deploying a smart contract](https://docs.alchemy.com/alchemy/tutorials/hello-world-smart-contract), and part 2 [interacting with your smart contract](https://docs.alchemy.com/alchemy/tutorials/hello-world-smart-contract/interacting-with-a-smart-contract) prior to starting part 3 below.
```
Comment (minimalsm) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:24]: Note: we'll should probably change this to the ethereum.org version once this is live

**Review by minimalsm — CHANGES_REQUESTED** [review]
Hey @nstrike2 thanks for getting changes in on these. Just doing another quick sweep on markdown/best practices and we can get to double-checking content 😀 thanks for your patience and persistence ❤️

**Comment by nstrike2** [comment]
> Note: we'll should probably change this to the ethereum.org version once this is live

Will do!

**Comment by nstrike2** [comment]
> Hey @nstrike2 thanks for getting changes in on these. Just doing another quick sweep on markdown/best practices and we can get to double-checking content 😀 thanks for your patience and persistence ❤️

You're the best @minimalsm 🙏 — just fixed your changes

**Comment by nstrike2** [comment]
@minimalsm wanted to follow up to get these all pushed!

**Comment by nstrike2** [comment]
@minimalsm can we merge this? I'll be leaving Alchemy in 3 days to go back to school and then can't do much from here to continue on this after the holidays, so would love to have these be mutually beneficial to both you and us! Have made all the changes — should be ready to go.

**Comment by nstrike2** [comment]
Following up @minimalsm !

**Comment by nstrike2** [comment]
> # [Gatsby Cloud](https://www.gatsbyjs.com) Build Report
> 🚩 Your build failed. [See the build logs here](https://gatsbyjs.com/dashboard/5e02f963-3fe5-4e37-a1e7-ac05f7753a86/sites/dc7c4340-a8e9-44ed-b7be-883c70d2f02d/builds/d1e39ff5-1376-4140-a916-485c609606d0/details)

@minimalsm not super sure why the build is failing, can't see the logs, as I don't have access (see screenshot) — would love some clarity here!

Screen Shot 2021-12-30 at 10 00 00 AM

File: src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+Successfully verified 1 contract(s).
+```
+
+### Step 4: Check out your smart contract on Etherscan! {#check-out-your-smart-contract-on-etherscan!}
+
+When you navigate to the link provided in your terminal, you should be able to see your smart contract code and ABI published on Etherscan!
+
+**Wahooo - you did it champ! Now anyone can call or write to your smart contract! We can't wait to see what you build next!** 🎉
+
+Once you complete this tutorial, let us know how your experience was or if you have any feedback by tagging us on Twitter [@alchemyplatform](https://twitter.com/AlchemyPlatform)!
```
Comment (sukruarc) [inline src/content/developers/tutorials/submitting-and-verifying-your-smart-contract-to-etherscan/index.md:250]: Eth

**Review by sukruarc — CHANGES_REQUESTED** [review]

**Comment by minimalsm** [comment]
Closing as per [this comment](https://github.com/ethereum/ethereum-org-website/pull/4284#issuecomment-1030104270)

**Pull request closed without merging**